volttron/platform/agent/base_tagging.py

```diff
@@ -0,0 +1,1044 @@
+- :py:meth:`BaseTaggingService.setup`
+- :py:meth:`BaseTaggingService.load_valid_tags`
+- :py:meth:`BaseTaggingService.load_tag_refs`
+- :py:meth:`BaseTaggingService.query_categories`
+- :py:meth:`BaseTaggingService.query_tags_by_category`
+- :py:meth:`BaseTaggingService.query_tags_by_topic`
+- :py:meth:`BaseTaggingService.query_topics_by_tags`
+- :py:meth:`BaseTaggingService.insert_topic_tags`
+
+On start calls the following methods
+- :py:meth:`BaseTaggingService.setup`
+- :py:meth:`BaseTaggingService.load_valid_tags`
+- :py:meth:`BaseTaggingService.load_tag_refs`
+
+Querying for topics based on tags
+---------------------------------
+Base tagging service provides a parser to parse query
+condition for querying topics based on tags. Please see documentation of
+- :py:meth:`BaseTaggingService.get_topics_by_tags` for syntax definition of
+query
+
+"""
+
+from __future__ import absolute_import, print_function
+
+import logging
+
+from abc import abstractmethod
+
+from volttron.platform.agent.known_identities import (PLATFORM_HISTORIAN)
+from volttron.platform.agent.utils import fix_sqlite3_datetime
+from volttron.platform.vip.agent import *
+from volttron.platform.vip.agent.errors import Unreachable
+
+try:
+    import ujson
+    def dumps(data):
+        return ujson.dumps(data, double_precision=15)
+    def loads(data_string):
+        return ujson.loads(data_string, precise_float=True)
+except ImportError:
+    from zmq.utils.jsonapi import dumps, loads
+
+_log = logging.getLogger(__name__)
+
+# Register a better datetime parser in sqlite3.
+fix_sqlite3_datetime()
+
+
+class BaseTaggingService(Agent):
+    """This is the base class for tagging service implementations. There can
+    be different implementations based on backend/data store used to persist 
+    the tag details
+    """
+
+    def __init__(self, resource_sub_dir='resources', **kwargs):
+        self.valid_tags = dict()
+        self.tag_refs = dict()
+        self.resource_sub_dir = "resources"
+        if resource_sub_dir:
+            self.resource_sub_dir = resource_sub_dir
+
+        super(BaseTaggingService, self).__init__(**kwargs)
+        _log.debug("Done init of base tagging service")
+
+    @Core.receiver("onstart")
+    def on_start(self, sender, **kwargs):
+        """
+        Called on start of agent. Calls the setup method
+        """
+        self.setup()
+        self.load_valid_tags()
+        self.load_tag_refs()
+
+    @abstractmethod
+    def setup(self):
+        """
+        Called on start of agent
+        Method to establish database connection, do any initial 
+        bootstrap necessary. Example - load master list of tags, units,
+        categories etc. into data store/memory
+        """
+        pass
+
+    @abstractmethod
+    def load_valid_tags(self):
+        """
+        Called right after setup to load a dictionary of valid tags. It
+        should load self.valid_tags with tag and type information
+
+        """
```

Comment:
add pass here

Reply:
true. should i just pass that as args or expect the function to return the valid tags

Reply:
I don't understand your response....